cvmfs/sqlitemem.h

```diff
@@ -0,0 +1,303 @@
+        return reinterpret_cast<AvailBlockCtl *>(base + link_next);
+      }
+      AvailBlockCtl *GetPrevPtr(char *base) {
+        return reinterpret_cast<AvailBlockCtl *>(base + link_prev);
+      }
+      int32_t ConvertToLink(char *base) {
+        return reinterpret_cast<char *>(this) - base;
+      }
+      int32_t size;  // always positive
+      int32_t link_next;
+      int32_t link_prev;
+    };
+
+    /**
+     * 8 bytes upper boundary of a free block.
+     */
+    struct AvailBlockTag {
+      explicit AvailBlockTag(int32_t s) : size(s), tag(kTagAvail) { }
+      int32_t size;
+      char padding[3];
+      char tag;
+    };
+
+    /**
+     * Lower boundary of a reserved block: a negative size to distinguish from
+     * the lower boundary of a free block.
+     */
+    struct ReservedBlockCtl {
+      explicit ReservedBlockCtl(int32_t s) : size(-s) {
+        char *base = reinterpret_cast<char *>(this);
+        *(base + s - 1) = kTagReserved;
+      }
+      int32_t size;  // always negative
+    };
+
+    /**
+     * Starts with the address of the MallocArena object followed by a
+     * head_avail_ block and ends with a -1 guard integer to mimic a reserved
+     * end block.  The arena is aligned at a multiple of kArenaSize.  Therefore,
+     * a pointer pointing into it can get the corresponding MallocArena object
+     * in O(1).
+     */
+    char *arena_;  ///< The memory block
+    /**
+     * Head of the list of available blocks.  Located at the beginning of the
+     * arena_.  The only "available" block with size 0 and with the upper tag
+     * field set to "reserved", so that it is not merged with another free'd
+     * block.
+     */
+    AvailBlockCtl *head_avail_;
+    AvailBlockCtl *rover_;  ///< The free block where the next search starts.
+    uint32_t no_reserved_;
+  };
+
+
+  /**
+   * A continues chunk of memory from which fixed-sized chunks are given as
+   * lookaside buffers to SQlite database connections.  A bitmap indicates used
+   * and free buffers.
+   * One arena serves 128 database connections and assuming less than 10000
+   * concurrently open catalogs, the number of mmap'd regions stays under 100.
+   */
+  class LookasideBufferArena {
+   public:
+    /**
+     * Accumulates to 4kB.
+     */
+    static const unsigned kBufferSize =
+      kLookasideSlotSize * kLookasideSlotsPerDb;
+
+    /**
+     * Has to be a multiple of the number of bits in the int type.  One arena
+     * can serve 128 concurrent sqlite database connections.
+     */
+    static const unsigned kBuffersPerArena = 128;
+    /**
+     * 512kB = 128 4kB buffers.
+     */
+    static const unsigned kArenaSize = kBuffersPerArena * kBufferSize;
+    /**
+     * Number of bitmap ints for 128 buffers.
+     */
+    static const unsigned kNoBitmaps = kBuffersPerArena / (sizeof(int) * 8);
+
+    LookasideBufferArena();
+    ~LookasideBufferArena();
+
+    void *GetBuffer();
+    void PutBuffer(void *buffer);
```

Comment:
`PutBuffer` (as well as `PutLookasideBuffer` and `PutMemory`) is freeing the memory chunk pointed to by `void*`, right? I find the naming unclear. Why not `Reserve***()` and  `Free***()`?